New scoring metrics for "cross_validation_across_time"

sianna338 opened the pull request:
The function **cross_validation_across_time** can now use different scoring algorithms to determine classifier performance. 
The parameter "metric" is added which accepts either a string keyword, refering to one of the scoring functions from sklearn.metrics, or alternatively a user-defined scoring function. Default is "accuracy".

sianna338 committed:
Update cross_validation_across_time to use different scoring functions

sianna338 committed:
cross_validation_across_time: allow self-defined metric function

skjerns commented on decoding.py:
```diff
@@ -154,7 +154,8 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
                                  tmin=-0.1, tmax=0.5, sfreq=100,
-                                 return_probas=False,
+                                 return_probas=False, metric='accuracy',
+                                 metric_kwargs={}, proba=True, 
```
We now have `proba` and `return_proba` - the seem to me as to do the same thing, no?

sianna338 replied:
you're right it doesnt make sense

skjerns commented on decoding.py:
```diff
@@ -247,7 +263,7 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
                 neg_x=neg_x,
                 clf=clf,
-                proba=True
+                proba=proba
```
shouldn't `proba` always be True? In what case should it be `False` ?

sianna338 replied:
```suggestion
                proba=return_proba
```

sianna338 replied:
some sklearn.metrics functions (like for example: "balanced_accuracy_score") expect predicted labels as input instead of the classifier probability per class. But there might be a more elegant way to solve this, since now the user has to check this requirement before selecting the scoring function

skjerns replied:
Indeed!

Have a look at `sklearn.metrics.get_scorer` and `sklearn.metrics.get_scorer` this can be used in this case quite elegantly to get a scoring method that works on all kinds of inputs and automatically does that afaik :)

sianna338 replied:
I checked out sklearn.metrics.get_scorer and make_scorer, but they don’t seem to fit this case.
get_scorer only returns predefined scorers and doesn’t return the expected input type (no response_method).
make_scorer lets you specify response_method (e.g. .predict_proba) but seems to work only with tools like cross_val_score or GridSearchCV.

skjerns commented on decoding.py:
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
 
         preds = np.argmax(probas, -1)
-
-        accuracy = (preds == test_y[:, None]).mean(axis=0)
+        
+        if metric == "accuracy": 
+            accuracy = (preds == test_y[:, None]).mean(axis=0)
+        else:
```
you can also do this using the `top_k_accuracy_score(k=1)` :) makes it a bit cleaner

skjerns commented on decoding.py:
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
+            # need to loop over timepoints 
+            accuracy = np.zeros(time_max)
+            for t in list(range(0, time_max)): 
+                accuracy[t] = func(data_y, all_probas[:,t], **kwargs)
```
maybe rename variable from accuracy to `score`, as it now can be any kind of result?

sianna338 replied:
```suggestion
            score= np.zeros(time_max)
            for t in list(range(0, time_max)): 
                score[t] = func(data_y, all_probas[:,t], **kwargs)
```

skjerns replied:
Don't just suggest things, implement them locally and push to the branch :D

sianna338 replied:
haha sorry, just trying out what happens when I click the buttons

skjerns commented on decoding.py:
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
+                kwargs = {}
+                for k, v in metric_kwargs.items():
+                    if k in sig.parameters:
+                        kwargs[k] = v
```
I would throw an error if the parameter is not part of the called function! otherwise we silently drop it and the user might not realize his kwarg had no effect.
```suggestion
            if metric_kwargs:
                sig = inspect.signature(func)
                missing_kwargs = set(metric_kwargs).difference(sig.parameters)
                if missing_kwargs:
                    raise ValueError(f'The following metric_kwargs were given but are not part of the function signature {missing_kwargs} ')
```

and then later simply pass metric_kwargs` instead of `kwargs` to the function call

skjerns commented on decoding.py:
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
+            # need to loop over timepoints 
+            accuracy = np.zeros(time_max)
+            for t in list(range(0, time_max)): 
+                accuracy[t] = func(data_y, all_probas[:,t], **kwargs)
```
```suggestion
                accuracy[t] = func(data_y, all_probas[:,t], **metric_kwargs)
```
after resolving the above comments

skjerns commented on decoding.py:
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
+                accuracy[t] = func(data_y, all_probas[:,t], **kwargs)
 
         # Create a temporary DataFrame for the current fold
         df_temp = pd.DataFrame(
```
here you should probably also change the metric name in the dataframe.

Additionally, adding the kwargs to the dataframe might be useful, but only add it if there were any given e.g.

`"metric_kwargs": str(metric_kwargs)`

sianna338 commented on decoding.py:
```diff
@@ -154,7 +154,8 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
                                  tmin=-0.1, tmax=0.5, sfreq=100,
-                                 return_probas=False,
+                                 return_probas=False, metric='accuracy',
+                                 metric_kwargs={}, proba=True, 
```
```suggestion
                                 metric_kwargs={}, 
```

sianna338 committed:
Update decoding.py

add error message when given metric_kwargs don't match the input arguments required by the scoring function

Co-authored-by: Simon Kern <14980558+skjerns@users.noreply.github.com>

sianna338 committed:
implemented some more suggested changes

input to function, default scoring function (top_1_accuracy)

sianna338 committed:
output variable naming fixed

sianna338 committed:
save metric_kwargs (optinal inputs for scoring function) in the final df

skjerns commented:
@sianna338 let me know when I should review again, and please comment and/or mark the conversations above as "resolved", so I know that it has been taken care of

sianna338 committed:
fix default case ("accuracy) for cross_validation_across_time

sianna338 committed:
try to automatically determine the input the performance metric function for the classifier needs

sianna338 committed:
implement way to determine which type of input the scoring function requires

sianna338 committed:
Merge branch 'skjerns:main' into main

skjerns commented:
/gemini review

skjerns commented:
/gemini summary

skjerns commented on .gitignore:
```diff
@@ -160,3 +160,5 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+decoding_test-1.py
+prepare_dat.py
```
@sianna338 

generally dont add them to the gitignore, but rather just dont add them to the commit instead! but shuold be no problem

skjerns commented on decoding.py:
```diff
@@ -18,12 +18,14 @@
 import pandas as pd
 import json
 import mne
+import misc
+import numpy as np
```
@sianna338 please fix

skjerns commented on decoding.py:
```diff
@@ -234,39 +245,84 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
+                raise TypeError("metric must be 'accuracy', a sklearn.metrics name (str), or a callable.")
+                        
+            # Determine if metric function expects probabilities or predictions. 
+            sig = inspect.signature(func)
+            inputs_names = list(sig.parameters)
+        
+            second_param = inputs_names[1]
+            prob_indicators = ['y_score', 'probas_pred', 'y_proba']  
+            pred_indicators = ['y_pred', 'labels_pred']
+            if any(indicator in second_param for indicator in prob_indicators):
+                needs_probas=True
+            elif any(indicator in second_param for indicator in pred_indicators):
+                needs_probas=False
+            else:
+                print("determining response method not possible")
```
you can get the needed response method by using the scorers that are returned from `get_scorer`

e.g.

```python
if isinstance(metric, str):
    # assume that it is a sklearn.metric scorer
    scorer = sklearn.metrics.get_scorer()
    # dont use scorer directly, but extract methods from it
    response_method = scorer._response_method
    metric_func= scorer._score_func
else:
    # simply assume the custom function will take predict_proba
    # if not, the user can simply write a wrapper around his function
    # that converts proba to class labels.
    response_method = 'predict_proba'
    metric_func= metric
```

then later you can write something like

```python
if response_method=='predict':
    ... code to use with labels
    metric_func(preds, y_true)
elif response_method=='predict_proba':
    ... code that uses probabilities to calculate
    metric_func(probabilities, y_true)
```

skjerns commented on decoding.py:
```diff
@@ -234,39 +245,84 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
+        all_results[idxs_test] = results_swp
+        
+        # convert probabilities to prediction if necessary 
+        if not needs_probas:
+           preds_idx = np.argmax(all_results, axis=2)              
+           preds_lbl = labels[preds_idx] 
```
move this down inside the other `else` clause, so it is where it's actually needed

sianna338 committed:
Merge branch 'skjerns:main' into main
